server/channels/jobs/s3_path_migration/s3_path_migration.go

```diff
@@ -0,0 +1,276 @@
+				mlog.Error("Worker: Failed to mark job as pending",
+					mlog.String("workername", worker.name),
+					mlog.String("job_id", job.Id),
+					mlog.Err(err))
+			}
+			return
+		case <-time.After(timeBetweenBatches):
+			var files []*model.FileForIndexing
+			tries := 0
+			for files == nil {
+				var err error
+				// Iterate through the rows in fileInfo, starting from the metadata ID
+				files, err = worker.store.FileInfo().GetFilesBatchForIndexing(int64(createAtOffset), offset, false, pageSize)
+				if err != nil {
+					if tries > 3 {
+						mlog.Error("Worker: Failed to get files after multiple retries. Exiting")
+						worker.setJobError(job, model.NewAppError("DoJob", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+						return
+					}
+					mlog.Warn("Failed to get file info for s3 migration. Retrying .. ", mlog.Err(err))
+
+					// Wait a bit before trying again.
+					time.Sleep(15 * time.Second)
+				}
+
+				tries++
+			}
+
+			if len(files) == 0 {
+				mlog.Info("S3PathMigrationWorker: Job is complete", mlog.String("worker", worker.name), mlog.String("job_id", job.Id))
+				worker.setJobSuccess(job)
+				worker.markAsComplete()
+				return
+			}
+
+			// Take batches of `pageSize`
+			for _, f := range files {
+				// We do not fail the job if a single image failed to encode.
+				if f.Path != "" {
+					if err := worker.fileBackend.DecodeFilePathIfNeeded(f.Path); err != nil {
+						mlog.Warn("Failed to encode S3 file path", mlog.String("path", f.Path), mlog.Err(err))
+					}
+				}
+				if f.PreviewPath != "" {
+					if err := worker.fileBackend.DecodeFilePathIfNeeded(f.PreviewPath); err != nil {
+						mlog.Warn("Failed to encode S3 file path", mlog.String("path", f.PreviewPath), mlog.Err(err))
+					}
+				}
+				if f.ThumbnailPath != "" {
+					if err := worker.fileBackend.DecodeFilePathIfNeeded(f.ThumbnailPath); err != nil {
+						mlog.Warn("Failed to encode S3 file path", mlog.String("path", f.ThumbnailPath), mlog.Err(err))
+					}
+				}
+			}
+
+			// Work on each batch and save the batch starting ID in metadata
+			lastFile := files[len(files)-1]
+			offset = lastFile.Id
+			createAtOffset = int(lastFile.CreateAt)
+
+			if job.Data == nil {
+				job.Data = make(model.StringMap)
+			}
+			job.Data["start_file_id"] = offset
+			job.Data["start_create_at"] = strconv.Itoa(createAtOffset)
+			doneCount += len(files)
+			job.Data["done_file_count"] = strconv.Itoa(doneCount)
+			job.Data["total_file_count"] = strconv.Itoa(totalCount)
+
+			worker.jobServer.SetJobProgress(job, int64((doneCount+len(files))/totalCount))
+		}
+	}
+}
+
+func (worker *S3PathMigrationWorker) markAsComplete() {
+	system := model.System{
+		Name:  model.MigrationKeyS3Path,
+		Value: "true",
+	}
+
+	if err := worker.jobServer.Store.System().Save(&system); err != nil {
+		mlog.Error("Worker: Failed to mark s3 path migration as completed.", mlog.String("workername", worker.name), mlog.Err(err))
```

Comment:
Shall we set job error here as well?

Reply:
Actually not here, because this happens _after_ we have already marked the job as success. To make this 100% fool proof, we'd need to do both in a single transaction. But that needs a lot more refactoring to allow methods from different store interfaces to run in a single transaction. I am going to let this one be with a comment to highlight the edge case.